## Week 9 Exercises

###<font color=green> Replace the name below with **your name**. For full credit, your total needs to add up to 50 points.

I highly recommend reading the accompanying pdf for programming notes and other important details.

# Kavin Kannan

In [8]:
import numpy as np
import matplotlib.pyplot as plt
import scipy
from scipy import linalg

1. Vector Dot and Cross Products (10 pts) Define two vectors in
the plane: u = [2, 0], v = [3, 4]. Calculate their dot product by hand,
state the result in a markdown cell. Then create these two vectors as
numpy arrays and show that u@v gives the same result. Repeat the
calculation and numerical computation for the cross product.

u = [2, 0] and v = [3, 4] Thus, the dot product is 2\*3 + 0\*4 = 6

In [2]:
u = np.array([2, 0], float)
v = np.array([3, 4], float)
print(u@v)

6.0


u = [2, 0] and v = [3, 4] Thus, the cross product is 2\*3 - 0\*4 = 8

In [3]:
print(np.cross(u,v))

8.0


/var/folders/j1/9bbht8xj16v37v43stsppv3c0000gn/T/ipykernel_16662/2181223691.py:1: DeprecationWarning: Arrays of 2-dimensional vectors are deprecated. Use arrays of 3-dimensional vectors instead. (deprecated in NumPy 2.0)
  print(np.cross(u,v))


2. Matrix multiplication (10 pts) Here are two matrices:

A =


1 3
2 4
5 6

 , B =

2 1 4
3 2 1 

Create these two matrices in Python, then print out A, B, and the
two matrix products A · B and B · A. (Hint: If you do something like
print(f’A =\n{aa}’) the matrix will print out nicely.) Calculate by
hand at least one element of A · B and one element of B · A and check
that these agree with your Python calculation.

In [4]:
aa = np.array([[1,2,5],[3,4,6]],float)
bb = np.array([[2,1],[4,3],[2,1]],float)
print('A =\n',aa)   
print('B =\n',bb)

A =
 [[1. 2. 5.]
 [3. 4. 6.]]
B =
 [[2. 1.]
 [4. 3.]
 [2. 1.]]


A * B will have a top left value of 1 * 2 + 2 * 4 + 5 * 2 = 20

In [6]:
print(aa@bb)

[[20. 12.]
 [34. 21.]]


B * A will have a top left value of 2 * 1 + 1 * 3 = 5

In [7]:
print(bb@aa)

[[ 5.  8. 16.]
 [13. 20. 38.]
 [ 5.  8. 16.]]


3. Inverse of a matrix (10 pts) Consider the matrix

C =


1 3 2
2 4 1
5 6 5



Have Python print out C, C−1

, and C−1

· C. Is the last of these what

you expect?

In [16]:
cc = np.array([[1,3,2],[2,4,1],[5,6,5]], float)
print('C =\n',cc)
ccinverse = linalg.inv(cc)
print('\nC inverse =\n',ccinverse)
print('\nC inverse * C =\n',ccinverse@cc)

C =
 [[1. 3. 2.]
 [2. 4. 1.]
 [5. 6. 5.]]

C inverse =
 [[-0.82352941  0.17647059  0.29411765]
 [ 0.29411765  0.29411765 -0.17647059]
 [ 0.47058824 -0.52941176  0.11764706]]

C inverse * C =
 [[ 1.00000000e+00  2.22044605e-16  2.22044605e-16]
 [-8.32667268e-17  1.00000000e+00 -8.32667268e-17]
 [ 2.22044605e-16  0.00000000e+00  1.00000000e+00]]


No the last one is not what I was expecting because it should return the identity matrix, however, it returns something very similar but with many small numbers as 0 and lots of 0s after the ones on the diagonal.

Solve a system of linear equations (10 pts) Use Python to solve
the following system of linear equations:
x + 4y + z = −4,
4x − y − z = 3,
2x + y + 3z = 7.

for the three unknowns x, y and z. Do this two different ways: (a) by
multiplying the inverse of the matrix of coefficients by the RHS vector,
and (b) using linalg.solve (which is more accurate in general), and
check that both methods agree. Plug the solution for x, y, z into one
of the original equations to check that it is satisfied.

In [21]:
m = np.array([[1,4,1],[4,-1,-1],[2,1,3]], float)
k = np.array([-4,3,7], float)
minverse = linalg.inv(m)
x = minverse@k

print(f'x = {x[0]}, y = {x[1]}, z = {x[2]}')

x = 0.8846153846153846, y = -1.8076923076923077, z = 2.346153846153846


In [ ]:
x = linalg.solve(m,k)

print(f'x = {x[0]}, y = {x[1]}, z = {x[2]}')

x = 0.8846153846153845, y = -1.8076923076923075, z = 2.3461538461538454


0.88461 + 4 * (-1.80769) + 2.34615 = -3.99999 = -4 = -4

Eigenvectors and eigenvalues of a real, symmetric matrix.(10
pts) Consider the following real, symmetric matrix

Sx =




0
√
3/2 0 0
√
3/2 0 1 0
0 1 0 √
3/2

0 0 √
3/2 0




.

(a) Compute and print out this matrix.
(b) Compute and print out the four eigenvectors and eigenvalues for
this matrix.
(c) Multiply the matrix by one of the eigenvectors and show that you
get what is expected.

In [33]:
ss = np.array([[0,np.sqrt(3)/2,0,0], [np.sqrt(3)/2,0,1,0], [0,1,0,np.sqrt(3)/2], [0,0,np.sqrt(3)/2,0]], float)
print('S =\n',ss)

S =
 [[0.        0.8660254 0.        0.       ]
 [0.8660254 0.        1.        0.       ]
 [0.        1.        0.        0.8660254]
 [0.        0.        0.8660254 0.       ]]


In [34]:
evals, evecs = linalg.eigh(ss)
print(evals)
print('\n',evecs)

[-1.5 -0.5  0.5  1.5]

 [[-0.35355339  0.61237244  0.61237244  0.35355339]
 [ 0.61237244 -0.35355339  0.35355339  0.61237244]
 [-0.61237244 -0.35355339 -0.35355339  0.61237244]
 [ 0.35355339  0.61237244 -0.61237244  0.35355339]]


In [36]:
v1 = evecs[:, 0]
print(ss@v1)

[ 0.53033009 -0.91855865  0.91855865 -0.53033009]


This result is a scale factor of -1.5 applied to the first eigenvector which is one of our eigenvalues so we're good to go! :)